Minor refactoring and add more tests

almarklein opened the pull request:
The change is smaller than the diff seems to suggest:
* Moved some code around to get rid of the mixin class. The idea of that class was to be able unit-test it better, but I realized that it does not really matter.
* Add more tests to validate the behavior of the widget.
* This incidentally also makes the autodoc entry easier.
* ... and the MRO cleaner :)

djhoese commented on docs/reference.rst:
```diff
@@ -6,7 +6,7 @@ RemoteFrameBuffer class
 
 .. autoclass:: jupyter_rfb.RemoteFrameBuffer
-    :members: request_draw, get_frame, handle_event, reset_stats, get_stats
+    :members:
```
I'm not sure I brought this up yet, but what about `sphinxcontrib.apidoc` to auto-generate the API files? Or is it small enough that that's overkill.

almarklein replied:
There's just two autodoc entries, so I'd consider it overkill indeed :)

djhoese commented on jupyter_rfb/widget.py:
```diff
@@ -23,22 +23,143 @@
-    track of stats. By making this a mixin, it is much easier to unit-test.
+@ipywidgets.register
+class RemoteFrameBuffer(ipywidgets.DOMWidget):
+    """A widget that shows a remote frame buffer.
+    Subclass of `ipywidgets.DOMWidget <https://ipywidgets.readthedocs.io>`_.
```
Have you double checked that this renders properly? Usually flake8/sphinx wants a blank line between first line and everything after.

djhoese replied:
```suggestion
    """A widget that shows a remote frame buffer.
    
    Subclass of `ipywidgets.DOMWidget <https://ipywidgets.readthedocs.io>`_.
```

almarklein replied:
Yes:
![image](https://user-images.githubusercontent.com/3015475/127400217-a11e2f9f-ae7d-403a-a902-65fdb2763028.png)

Now that you mention it, I also recall that to be a thing. But neither flake nor Sphinx complains, and I like how the two lines become a single line in the docs (instead of two very short paragraphs) , so my preference would be to keep it like this.

djhoese replied:
:thinking: Well the first line is supposed to be separate. It is the "subject line" of the docstring and the rest is the "body" of the docstring.

I think you are missing the flake8-docstring plugin package. It fails on my machine for this branch:

```
$ flake8 jupyter_rfb/widget.py
jupyter_rfb/widget.py:28:1: D205 1 blank line required between summary line and description
jupyter_rfb/widget.py:68:1: D107 Missing docstring in __init__
jupyter_rfb/widget.py:81:1: D205 1 blank line required between summary line and description
jupyter_rfb/widget.py:81:1: D400 First line should end with a period
jupyter_rfb/widget.py:81:1: D402 First line should not be the function's "signature"
jupyter_rfb/widget.py:100:1: D205 1 blank line required between summary line and description
jupyter_rfb/widget.py:100:1: D400 First line should end with a period
jupyter_rfb/widget.py:165:1: D205 1 blank line required between summary line and description
jupyter_rfb/widget.py:182:1: D205 1 blank line required between summary line and description
jupyter_rfb/widget.py:182:1: D400 First line should end with a period
jupyter_rfb/widget.py:229:1: D205 1 blank line required between summary line and description
jupyter_rfb/widget.py:229:1: D400 First line should end with a period
jupyter_rfb/widget.py:229:1: D401 First line should be in imperative mood; try rephrasing
jupyter_rfb/widget.py:236:1: D205 1 blank line required between summary line and description
jupyter_rfb/widget.py:236:1: D400 First line should end with a period
jupyter_rfb/widget.py:236:1: D401 First line should be in imperative mood; try rephrasing
```

djhoese commented on tests/test_widget.py:
```diff
@@ -1,11 +1,18 @@
+"""
+Tests the RemoteFrameBuffer widget class.
```
```suggestion
"""Tests the RemoteFrameBuffer widget class.
```

djhoese commented on tests/test_widget.py:
```diff
@@ -161,6 +171,88 @@ def test_framesender_3():
 if __name__ == "__main__":
-    test_framesender_1()
-    test_framesender_3()
+    test_widget_frames_and_stats_1()
+    test_widget_frames_and_stats_3()
+    test_get_frame_can_be_none()
+    test_widget_traits()
+    test_widget_default_get_frame()
+    test_requesting_draws()
+    test_automatic_events()
```
Isn't there like a `pytest.main()` or something that could be run here instead?

almarklein replied:
meeh, I just removed these. I can run pytest from the command-line.

djhoese reviewed (changes requested):
Hard to tell what's changed with git's diff. I got a little picky with some stuff, but :shrug: you asked for a review.

djhoese commented:
Ok be prepared to hate me. I added additional dependencies to the flake8 hook in the pre-commit config. This makes it warn about a lot more things. I also fixed all the issues in widget.py *except* I updated setup.cfg to ignore two things:

1. Ignore no docstring on `__init__`. To me, these usually get redundant or pointless.
2. `jupyter_rfb/widget.py:130:13: W503 line break before binary operator`. It appears this is one thing black does that flake8 doesn't like. I ignored it but I probably could have extracted the individual conditions to clearly named variables but I didn't know what the code was doing so I just ignored it.

Feel free to remove the flake8 dependencies or clean up that condition on line 130, but not sure we should completely undo what flake8 is requesting.

almarklein commented:
I must admit that I feel some tools get more in the way than actually making things easier. I didn't even know the "each docstring should be a one-liner" was formalized in a pep. I wonder how many tools adopt it - ipywidgets does not ;)

That said, I am open to give it a try, and I do see the value of having to document each method (even in tests). I feel that only adding it to the commit-hook does not do much. (I've uninstalled the hook TBH, I've tried ... I really did, but at some point it worked against me and I removed it again.) So I added the flake8 extensions to CI, and added them to the dev notes.

almarklein commented:
If you're happy we can merge this. Then I plan on doing a release and announce on Twitter. We can do a second announcement once Vispy has a new release that makes use of it, with prettier pictures. I plan to have a look at some of the performance optimizations a few weeks later.

djhoese commented on README.md:
```diff
@@ -63,6 +63,10 @@ You then need to refresh the JupyterLab page when your javascript changes.
 
+To install developer tools:
+
+    $ pip install pytest black flake8 flake8-docstrings flake8-bugbear
```
In the future it may be nice to have `pip install -r dev_requirements.txt` or something like that. Don't worry about it now.

djhoese reviewed (approved):
Looks good to me. I had one small comment for the future that I put inline. Another future thing, is that the website could be built in CI and include the proper flags to error on warnings. This would make sure that all pull requests don't add sphinx/documentation issues. Not super important right now though.

djhoese commented:
I'll let you merge this. I would like to investigate the vispy installation issue I ran into yesterday some more. I can't run examples again and trying to reinstall doesn't do anything.

almarklein commented:
> I can't run examples again and trying to reinstall doesn't do anything. Feel free to make announcements.

Ugh ... I remember not being able to uninstall an extension because the extension was in a broken state, and then manually removing a directory in the ipython configuration. Good luck.

djhoese commented:
Have you looked at adding the package to conda-forge?

almarklein commented:
No, not yet.

almarklein commented:
mmm, Binder seems to still use 0.1.1, not 0.1.2 that I just released. Not a big problem, but I assumed it would rebuild.

djhoese commented:
My binder session rebuilt and is using 0.1.2
